src/idmap/src/lib.rs

```diff
@@ -0,0 +1,216 @@
+    IDMAP_NOT_IMPLEMENTED,
+    IDMAP_ERROR,
+    IDMAP_OUT_OF_MEMORY,
+    IDMAP_NO_DOMAIN,
+    IDMAP_CONTEXT_INVALID,
+    IDMAP_SID_INVALID,
+    IDMAP_SID_UNKNOWN,
+    IDMAP_NO_RANGE,
+    IDMAP_BUILTIN_SID,
+    IDMAP_OUT_OF_SLICES,
+    IDMAP_COLLISION,
+    IDMAP_EXTERNAL,
+    IDMAP_NAME_UNKNOWN,
+    IDMAP_ERR_LAST,
+}
+
+fn map_err(e: ffi::idmap_error_code) -> IdmapError {
+    match e {
+        ffi::idmap_error_code_IDMAP_SUCCESS => IdmapError::IDMAP_SUCCESS,
+        ffi::idmap_error_code_IDMAP_NOT_IMPLEMENTED => IdmapError::IDMAP_NOT_IMPLEMENTED,
+        ffi::idmap_error_code_IDMAP_ERROR => IdmapError::IDMAP_ERROR,
+        ffi::idmap_error_code_IDMAP_OUT_OF_MEMORY => IdmapError::IDMAP_OUT_OF_MEMORY,
+        ffi::idmap_error_code_IDMAP_NO_DOMAIN => IdmapError::IDMAP_NO_DOMAIN,
+        ffi::idmap_error_code_IDMAP_CONTEXT_INVALID => IdmapError::IDMAP_CONTEXT_INVALID,
+        ffi::idmap_error_code_IDMAP_SID_INVALID => IdmapError::IDMAP_SID_INVALID,
+        ffi::idmap_error_code_IDMAP_SID_UNKNOWN => IdmapError::IDMAP_SID_UNKNOWN,
+        ffi::idmap_error_code_IDMAP_NO_RANGE => IdmapError::IDMAP_NO_RANGE,
+        ffi::idmap_error_code_IDMAP_BUILTIN_SID => IdmapError::IDMAP_BUILTIN_SID,
+        ffi::idmap_error_code_IDMAP_OUT_OF_SLICES => IdmapError::IDMAP_OUT_OF_SLICES,
+        ffi::idmap_error_code_IDMAP_COLLISION => IdmapError::IDMAP_COLLISION,
+        ffi::idmap_error_code_IDMAP_EXTERNAL => IdmapError::IDMAP_EXTERNAL,
+        ffi::idmap_error_code_IDMAP_NAME_UNKNOWN => IdmapError::IDMAP_NAME_UNKNOWN,
+        ffi::idmap_error_code_IDMAP_ERR_LAST => IdmapError::IDMAP_ERR_LAST,
+        _ => {
+            error!("Unknown error code '{}'", e);
+            IdmapError::IDMAP_ERROR
+        }
+    }
+}
+
+fn object_id_to_sid(object_id: &Uuid) -> Result<String, IdmapError> {
+    let bytes_array = object_id.as_bytes();
+    let s_bytes_array = [
+        bytes_array[6],
+        bytes_array[7],
+        bytes_array[4],
+        bytes_array[5],
+    ];
+
+    Ok(format!(
+        "S-1-12-1-{}-{}-{}-{}",
+        u32::from_be_bytes(
+            bytes_array[0..4]
+                .try_into()
+                .map_err(|_| IdmapError::IDMAP_SID_INVALID)?
+        ),
+        u32::from_be_bytes(
+            s_bytes_array
+                .try_into()
+                .map_err(|_| IdmapError::IDMAP_SID_INVALID)?
+        ),
+        u32::from_le_bytes(
+            bytes_array[8..12]
+                .try_into()
+                .map_err(|_| IdmapError::IDMAP_SID_INVALID)?
+        ),
+        u32::from_le_bytes(
+            bytes_array[12..]
+                .try_into()
+                .map_err(|_| IdmapError::IDMAP_SID_INVALID)?
+        )
+    ))
+}
+
+pub const DEFAULT_IDMAP_RANGE: (u32, u32) = (200000, 2000200000);
+
+// The ctx is behind a read/write lock to make it 'safer' to Send/Sync.
+// Granted, dereferencing a raw pointer is still inherently unsafe.
+pub struct SssIdmap {
+    ctx: RwLock<*mut ffi::sss_idmap_ctx>,
```

Comment:
Right here is where I mean. That raw pointer was preventing me from calling this library from the idprovider.

Reply:
Hrm, still interested in your opinion, but it looks like this whole approach is just not going to work. SSSD idmapping depends heavily on SIDs. There are no SIDs in AAD. I can fake a SID for users/groups, but not for the domain. Then SID to DOM_SID matching doesn't work anyway. Sigh

Reply:
I was actually going to caution you out of linking to SSS as a whole given it's history of "quality". SID mapping to an ID should be an algorithm we can easily copy if needed anyway. 

In this case, I think we need some other way to generate stable uids. If each user has a uuid you could take the last 32 bits as the uid/gid number perhaps? We do this in kanidm, but you do quickly approach the birthday problem ...

Reply:
> I'm assuming because it caused the async to fail because you were no longer send/sync?

Yeah, exactly.

Reply:
> In this case, I think we need some other way to generate stable uids. If each user has a uuid you could take the last 32 bits as the uid/gid number perhaps? We do this in kanidm, but you do quickly approach the birthday problem ...

This is something I thought of doing. I had hoped to exactly mirror the behavior of either SSSD or Winbind, but neither will really work well with these IDs vs SIDs. So I think we'll need to come up with our own similar approach.

Reply:
Do we know how windows turns an azure ad uuid into a SID? That could be what we need to re-create actually.